File: Algorithm/二分查找/有效的完全平方数367/main.go
/*
 * File              : main.go
 * Author            : Andy963
 * Created time      : 2024-12-07 22:03:05
 * Last Modified by  : Andy963
 * Last Modified time: 2024-12-08 08:07:45
 */

package main

import "fmt"

func solve(num int) bool {
	if (num == 1) {
		return true
	}
	left, right := 0, num
	for (left <= right) {
		mid := left + (right-left) / 2
		square := mid * mid
		if num > square{
			left = mid + 1
		} else if num < square {
			right = mid - 1
		} else {
			return true
		}
	}
	return false
}

func main() {
	num := 5 
	rs := solve(num)
	if rs == true {
		fmt.Println(true)
	}else{
		fmt.Println(false)
	}
}

File: Algorithm/二分查找/有效的完全平方数367/main.py
#!/usr/bin/env python3
# -*- encoding=utf8 -*-

'''
File              : main.py
Author            : Andy963
Created time      : 2024-12-07 21:57:32
Last Modified by  : Andy963
Last Modified time: 2024-12-07 22:02:50
'''
import pytest


class Solution:
    def solve(self, num:int):
        left, right = 0, num
        if num == 1:
            return True
        while left <= right:
            mid = left +(right - left) // 2
            if num / mid > mid:
                left = mid + 1
            elif num / mid < mid:
                right = mid - 1
            else:
                return True
        return False

sl = Solution()

def test_case1():
    num = 16
    expected = True
    assert sl.solve(num) == expected


def test_case2():
    num = 14
    expected = False
    assert sl.solve(num) == expected


def test_case3():
    num = 1
    expected = True
    assert sl.solve(num) == expected


if __name__ == '__main__':
    pytest.main(['-v','-s'])